gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+import org.eclipse.ditto.base.model.headers.DittoHeaders;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.CheckPermissions;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.CheckPermissionsResponse;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.ImmutablePermissionCheck;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.PermissionCheckWrapper;
+import org.eclipse.ditto.internal.utils.pekko.logging.DittoLogger;
+import org.eclipse.ditto.internal.utils.pekko.logging.DittoLoggerFactory;
+import org.eclipse.ditto.json.JsonFieldSelector;
+import org.eclipse.ditto.policies.model.PolicyId;
+import org.eclipse.ditto.policies.model.ResourceKey;
+import org.eclipse.ditto.policies.model.ResourcePermissionFactory;
+import org.eclipse.ditto.policies.model.ResourcePermissions;
+import org.eclipse.ditto.policies.api.commands.sudo.CheckPolicyPermissions;
+import org.eclipse.ditto.policies.api.commands.sudo.CheckPolicyPermissionsResponse;
+import org.eclipse.ditto.things.api.commands.sudo.SudoRetrieveThing;
+import org.eclipse.ditto.things.api.commands.sudo.SudoRetrieveThingResponse;
+import org.eclipse.ditto.things.model.Thing;
+import org.eclipse.ditto.things.model.ThingId;
+
+/**
+ * Actor that handles permission checks for resources and entities.
+ * <p>
+ * This actor receives {@link CheckPermissions} messages, processes permission checks by
+ * aggregating results based on policies and entities, and communicates with the Ditto Edge
+ * and Things API to retrieve policy data and permissions.
+ * It responds to the sender with a {@link CheckPermissionsResponse} after evaluating the
+ * permission checks.
+ * <p>
+ * The actor forwards requests to the edge command forwarder to retrieve policies and permission
+ * results, handling various entity and policy resources.
+ *
+ * @since 3.7.0
+ */
+final class CheckPermissionsActor extends AbstractActor {
+
+    private final DittoLogger logger = DittoLoggerFactory.getLogger(CheckPermissionsActor.class);
+    private final ActorRef edgeCommandForwarder;
+    private final ActorRef sender;
+    private final Duration defaultAskTimeout;
+
+    /**
+     * The actor name for this class.
+     */
+    public static final String ACTOR_NAME = "checkPermissionsActor";
+
+    /**
+     * Constructor to initialize the actor.
+     *
+     * @param edgeCommandForwarder the actor responsible for forwarding commands to edge services.
+     * @param sender the actor reference to the original sender.
+     * @param defaultTimeout the default timeout for async operations.
+     */
+    private CheckPermissionsActor(final ActorRef edgeCommandForwarder, final ActorRef sender,
+            final Duration defaultTimeout) {
+        this.edgeCommandForwarder = edgeCommandForwarder;
+        this.sender = sender;
+        this.defaultAskTimeout = defaultTimeout;
```

Comment:
We should add a "receive timeout" to the actor:
```
getContext().setReceiveTimeout(defaultTimeout);
```

If the actor does not receive any message within that timeout, it will get a `ReceiveTimeout` message, which you can handle in `createReceive` below:
```
.match(ReceiveTimeout.class, receiveTimeout -> this.handleReceiveTimeout())
```

That way, we can eventually shutdown the actor, in case something goes wrong.